### PR TITLE
Remove newlines from docstring signature

### PR DESCRIPTION
**Pull request by JeanElsner**
## Description
Addresses issue https://github.com/pybind/pybind11/issues/4734 by removing all newlines from function signature docstring.

## Suggested changelog entry:

```rst
Enforce single line docstring signatures.
```

### CODE REVIEW

File: include/pybind11/pybind11.h
Comment (rwgk) [inline include/pybind11/pybind11.h:465]: Could this be more targeted?

If the newlines come in through default values, could this be a better place to strip them out (currently line 427)?

```
signature += rec->args[arg_index].descr;
```

I don't think just stripping out newlines is safe in the general case. That might join two "words" into one, which could be super misleading (you have assume that a human glancing through a pile of docstrings has little to no domain knowledge, and we don't want to confuse them). I think we need something more sophisticated, roughly, replace newlines with spaces, but squash repeats, and strip off leading and trailing spaces.

But what we need foremost: test cases, added under pybind11/tests. That's probably 99+% of the work needed here, but a strict requirement for a core tool like this one.

I recommend starting we one test case, work on the implementation, add another test case, refine the implementation, etc. until the test suite covers all corner cases.

Comment (JeanElsner) [reply include/pybind11/pybind11.h:465]: Thank you for the quick response. I expected as much, but wasn't quite sure where to even begin I'll start looking into it.

Comment (rwgk) [reply include/pybind11/pybind11.h]: You seem to have embraced doing the implementation first, which makes sense as long as it remains relatively simple.

If you push the implementation to the point that all existing tests pass, I could run Google's global testing to see if anything gets upset (I'd guess not, but I've been weirdly surprised before several times; changing docstrings generally gets me worried more than other things).

I.e. the idea is:

* Please let me know when you have all tests working here, but maybe hold off writing the tests.
* I run global testing (takes a lot of computing resources, but not a lot of my time).
* If there are no surprises you add the tests.

**Comment by rwgk** [comment]
Everything works here now. I'll try to run this through global testing asap.

**Comment by rwgk** [comment]
Good news: the global testing is fine (no surprises).

I think with added tests this PR will be good.

(For myself, in case I need to backtrack later, this is the internal ID for the global testing: OCL:546641582:BASE:546634977:1688902402537:16f03b9e)

**Review by antonkesy — APPROVED** [review]

**Comment by rwgk** [comment]
Did you find out already how to deal with this?

```
ModuleNotFoundError: No module named 'numpy'
```

See e.g. test_stl_binders.py:

```
def test_vector_buffer_numpy():
    np = pytest.importorskip("numpy")
```

**Comment by JeanElsner** [comment]
> Did you find out already how to deal with this?
> 
> ```
> ModuleNotFoundError: No module named 'numpy'
> ```

I don't quite understand why some of the checks are failing now. I'm only using Eigen types in `test_eigen_matrix` which already imports numpy at module level. My only other modifications are in `test_kwargs_and_defaults` and there I didn't use any Eigen/numpy types. Any support would be really appreciated.

**Comment by JeanElsner** [comment]
I could reproduce the error in Docker and it's a bit convoluted. It turns out that if I set the default value to an Eigen type, numpy is imported by the `pybind11_tests` CPython module (so before pytest.importorskip is executed in the Python test). If I remove the Eigen defaults or simply install numpy the tests run successfully.

Is there a way to check for numpy in C++ similar to pytest.importorskip? Otherwise I would remove those tests and only test with custom types (none of the existing numpy/Eigen tests cases test defaults either).

**Comment by JeanElsner** [comment]
I think it's simply [scipy.sparse](https://github.com/scipy/scipy/blob/0aa5aa9e78f81493f1cd9f6fc149a0bb856c2e11/scipy/sparse/_sputils.py#L6) imported in the Eigen code that in turn imports numpy.

**Comment by rwgk** [comment]
> I think it's simply [scipy.sparse](https://github.com/scipy/scipy/blob/0aa5aa9e78f81493f1cd9f6fc149a0bb856c2e11/scipy/sparse/_sputils.py#L6) imported in the Eigen code that in turn imports numpy.

That would be really odd: AFAIK scipy cannot be installed without installing numpy first. At least that seems highly likely for GitHub Actions jobs.

Jumping ahead:

I wonder if we could try importing numpy from C++ and `.def()` the new functions only if that succeeds.

I wouldn't be surprised if that runs into weird difficulties, but I'd try.

Another idea would be to add a `-DPYBIND11_TESTS_HAVE_NUMPY` via ci.yml somehow and test for that from C++. We don't need to cover a lot of platforms, just one would be a good start.

Giving up is another option, but similar to before, I'd try at least a little bit more.

**Comment by JeanElsner** [comment]
You are right about the scipy dependency of course. Actually, the tests succeed without scipy but only numpy installed. I will investigate further and try your suggestions, thanks!

**Comment by rwgk** [comment]
The two CI failures are unrelated: one is PR #4767, the other one a frequent flake (test_iostream).

File: include/pybind11/pybind11.h
Comment (rwgk) [inline include/pybind11/pybind11.h:55]: Could you please move this inside the function?

If it needs to stay outside, we'd need a more specific name (long is fine), but generally it works best in my experience to keep such variables in the context in which they are needed, and pull them out only if they are actually needed elsewhere.

File: tests/test_kwargs_and_defaults.py
Comment (rwgk) [inline tests/test_kwargs_and_defaults.py:28]: It's not "real" that a string repr comes out without quotes.
It doesn't really matter, but to not have something that looks too artificial (and gets people scratching their head until they figure it out), I'd try to emulate a real-world situation more closely, within reason.
In this case, I'd try to simply not use something that looks like a string.
Maybe: take an actual Eigen repr, simplify/shorten, then add in whitespace like you do here.

But there is also a more substantial question: **What if there is a str repr with newlines?** We don't want to change that.

Comment (JeanElsner) [reply tests/test_kwargs_and_defaults.py:28]: You're right, this looks confusingly like a string.

I'm not sure I understand the second part correctly. These changes only effect the repr within the automatically generated docstring signatures. The repr of objects outside of this context stay the same. If your concern is the removal of intentional newlines within a string default value in the human-readable documentation, I think the broken generated stubs and/or Sphinx documentation would be more confusing to users. Personally, it took me quite a while to figure out what was going wrong.

Alternatively, it might be possible to check the `argument_record` for string types and not apply the stripping function at the cost of incompatibility with the stub and documentation tools for this specific case.

Comment (rwgk) [reply tests/test_kwargs_and_defaults.py:28]: > If your concern is the removal of intentional newlines within a string default value in the human-readable documentation,

Yeah, that's where my mind was going.

> I think the broken generated stubs and/or Sphinx documentation would be more confusing to users.

I agree.

> Alternatively, it might be possible to check the `argument_record` for string types and not apply the stripping function

That sounds like opening a can of worms.

Quoting again:
> intentional newlines within a string default value

Could you please add a test case? — I'm assuming that's really easy and then we have something concrete to look at. And others will see that we thought about it.

IIUC with the current state of the PR the newline will get replaced with a space.

Wondering:

* How about inserting something like `<nl>` instead?

* Should we do that unconditionally (also for the Eigen case)?

* Or something like: if we see any quote (`'` or `"`) in the docstring, we use `<nl>`, otherwise we use a space?

(I don't want to turn this into a science project, but also not sweep this under the rug without thinking a least a little bit.)

Comment (JeanElsner) [reply tests/test_kwargs_and_defaults.py]: I like the idea of adding something like `<nl>`. It might look a little confusing for the Eigen case (see [here](https://jeanelsner.github.io/panda-py/panda_py.controllers.html#panda_py.controllers.CartesianImpedance) for a real-life documentation example).

Your last suggestion seems like a nice compromise to me. How would propose the "parsing" would look like? Simply checking for the presence of quotes, checking the first and last characters or something more sophisticated even?

Comment (rwgk) [reply tests/test_kwargs_and_defaults.py]: What I had in mind was simply `s.find('\'') || s.find('"')`

Since you like the `<nl>` idea: I'd give it ~5 minutes looking around for prior art (`<NL>`, `<newline>`, ...) and emulate that. But not more than a few minutes.

Comment (JeanElsner) [reply tests/test_kwargs_and_defaults.py]: I added a test case and it turns out string representations don't include the newlines anyway, but rather print `\n` literally.

File: include/pybind11/pybind11.h
Comment (rwgk) [inline include/pybind11/pybind11.h:62]: Could you please make this `std::strchr`?

Elsewhere we have `std::strcmp` consistently, and we already have `#include <cstring>`.

File: include/pybind11/pybind11.h
Comment (rwgk) [inline include/pybind11/pybind11.h:56]: Could you please sprinkle `\r`, `\f`, `\v` into the tests below?

What we want is that a test fails if any of the characters here is accidentally lost. (Imagine a silly typo slipping in while refactoring.)

**Review by rwgk — APPROVED** [review]
Thanks a lot for the making this a high quality PR!

**Comment by rwgk** [comment]
(FYI: waiting a bit for feedback from other maintainers)

**Comment by rwgk** [comment]
There was a failure I've not seen before:

docs/readthedocs.org:pybind11 — Read the Docs build failed!

https://readthedocs.org/projects/pybind11/builds/21602133/

Probably transient, because it's working for #4786. I'll close-open this PR to re-trigger GHA.

**Comment by rwgk** [comment]
I took a super quick look, this PR doesn't touch any rst or config files, there is no obvious reason why readthedocs would fail AFAICT.

**Comment by rwgk** [comment]
Trying GHA again without any changes in this PR.

**Comment by rwgk** [comment]
Failing on master already.

Strangely, we didn't have that error under #4786: https://readthedocs.org/projects/pybind11/builds/21599337/

Maybe it broke shortly after?

**Comment by rwgk** [comment]
Hi @JeanElsner I fixed the readthedocs problem (#4789). Could you please git merge master, git push again?

**Comment by rwgk** [comment]
No feedback from any other maintainer for 7 days (I sent two requests for feedback). Merging.

**Pull request merged**
